go/state/state_db.go

```diff
@@ -596,84 +597,86 @@ func clone(val *big.Int) *big.Int {
 	return res
 }
 
-func (s *stateDB) GetBalance(addr common.Address) *big.Int {
+func (s *stateDB) GetBalance(addr common.Address) amount.Amount {
 	// Check cache first.
 	if val, exists := s.balances[addr]; exists {
-		return clone(&val.current) // Do not hand out a pointer to the internal copy!
+		return val.current
 	}
 	// Since the value is not present, we need to fetch it from the store.
 	balance, err := s.state.GetBalance(addr)
 	if err != nil {
 		s.errors = append(s.errors, fmt.Errorf("failed to load balance for address %v: %w", addr, err))
-		return new(big.Int) // We need to return something that allows the VM to continue.
+		return amount.NewAmount() // We need to return something that allows the VM to continue.
 	}
-	res := balance.ToBigInt()
+	res := amount.NewAmountFromBytes(balance[:]...)
 	s.balances[addr] = &balanceValue{
-		original: res,
-		current:  *res,
+		original: &res,
+		current:  res,
 	}
-	return clone(res) // Do not hand out a pointer to the internal copy!
+	return res
 }
 
-func (s *stateDB) AddBalance(addr common.Address, diff *big.Int) {
+func (s *stateDB) AddBalance(addr common.Address, diff amount.Amount) {
 	s.createAccountIfNotExists(addr)
 
-	if diff == nil || diff.Sign() == 0 {
+	if diff.IsZero() {
 		s.emptyCandidates = append(s.emptyCandidates, addr)
 		return
 	}
-	if diff.Sign() < 0 {
-		s.SubBalance(addr, diff.Abs(diff))
-		return
-	}
 
 	oldValue := s.GetBalance(addr)
-	newValue := new(big.Int).Add(oldValue, diff)
+	newValue, overflow := amount.AddOverflow(oldValue, diff)
+
+	if overflow {
+		s.errors = append(s.errors, fmt.Errorf("failed to add balance for address %v: overflow", addr))
```

Comment:
this is dangerous, it will invalidate the whole carmen database. Since the overflow comes likely from a blockchain transaction, there is not way to recover, every re-run will end up in this same state. 

how is overflow dealt with in blockchains? I guess it is assumed it cannot happen.

Reply:
It is even impossible to happen. Total supply of FTM is ~3 billions. Max value of uint256 is 2^256 - 1.

Reply:
Yes, it should indeed not happen. The total supply is not even close to max value of uint256, so If one wallet held all tokens, it would still fit.

Reply:
This is an inherent problem of the `StateDB` interface we adapted from `geth`.

I think in our design we should deprecate and eventually remove the `SubBalance` and `AddBalance` functions and replace them by a `SetBalance`. This way, it is the API client's job to deal with potential overflows.

Reply:
yes, makes sense. But for this moment I guess we should not treat this situation as an error, right?

Reply:
It is already treated as an error. As it used to be stored as `big.Int`, it was not checked immediately in Add/Sub methods. Instead it was checked in `EndBlock` method when the `big.Int` was converted to `Balance` with [ToBalance](https://github.com/Fantom-foundation/Carmen/blob/46353a6b7b9ab551a557f897e1175b755c5d70cc/go/common/types.go#L190) method. If the [conversion](https://github.com/Fantom-foundation/Carmen/blob/46353a6b7b9ab551a557f897e1175b755c5d70cc/go/state/state_db.go#L1193) failed, it produced an error.

Reply:
thanks for explanation. We can let it be as it is now I guess